File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn, main) where

main = print(ageOn Earth 1000000000)

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

planetGetEarthDays :: Planet -> Float
planetGetEarthDays planet =
    case planet of
        Mercury -> 0.2408467
        Venus -> 0.61519726
        Earth -> 1
        Mars -> 1.8808158
        Jupiter -> 11.862615
        Saturn -> 29.447498
        Uranus -> 84.016846
        Neptune -> 164.79132

ageOn :: Planet -> Float -> Float
ageOn planet seconds = seconds / (planetGetEarthDays(planet) * 86400 * 365.25)

File: fsharp/space-age/SpaceAge.fs
module SpaceAge

type Planet = Earth | Mercury | Venus | Mars | Jupiter | Saturn | Uranus | Neptune

let getPlanetFactor (planet: Planet) = 
    let earthSeconds = 31557600.0 in
        match planet with
        | Earth -> earthSeconds
        | Mercury -> earthSeconds * 0.2408467
        | Venus -> earthSeconds * 0.61519726
        | Mars -> earthSeconds * 1.8808158
        | Jupiter -> earthSeconds * 11.862615
        | Saturn -> earthSeconds * 29.447498
        | Uranus -> earthSeconds * 84.016846
        | Neptune -> earthSeconds * 164.79132

let age (planet: Planet) (seconds: int64): float =
    let seconds = seconds |> float in 
        System.Math.Round (seconds / getPlanetFactor planet, 2)